### ISSUE TITLE
Errors pop up with definitions

### ISSUE THREAD

**Issue by ababen1** [coppolaemilio/dialogic#388]
## The problem

**Describe the bug**
when I try to run a timeline, i get these errors:
"invalid get index 'value' (on base: dictionary) "
"[Dialogic] Error opening default definitions file: 12"

**To Reproduce**
1. Go to definitions > add value 
![image](https://user-images.githubusercontent.com/80917990/124596524-0d3e8480-de6b-11eb-952f-6eea86a82a82.png)

2. make a timeline with the value, for example i made a timeline called "ababa" with these events
![image](https://user-images.githubusercontent.com/80917990/124596681-3bbc5f80-de6b-11eb-8411-cd5f92e47895.png)
 
3. try to run the timeline by adding it as a child of Node2D, like so:
![image](https://user-images.githubusercontent.com/80917990/124597025-9fdf2380-de6b-11eb-8dc4-15c3bd86859e.png)

4. the "hello" shows up fine, but then this error happens:
![image](https://user-images.githubusercontent.com/80917990/124597205-d87efd00-de6b-11eb-82ab-c695512d77e0.png)

5. close the game, go back to the editor. in addition to the above error the new value doesnt show up in the dropdown anymore
![image](https://user-images.githubusercontent.com/80917990/124597419-2562d380-de6c-11eb-9253-3a1af4392386.png)

this error pops up when i click on the value
![image](https://user-images.githubusercontent.com/80917990/124598815-b7b7a700-de6d-11eb-9b0c-3418ee58473b.png)

**Expected behavior**
I expect to see my value in the dialog or at least for the game to not crash i guess 

**System (please complete the following information):**
 - OS: Windows 10
 - Godot Version: 3.3.2.stable
 - Dialogic Version: 1.2.4

## Solutions

**Workaround**

What you tried to get the feature working:
I tried to delete the addon and all the files and install it again and it didnt work

**Possible fixes**

Any idea to fix the issue or a link to the line of code that might be the cause for this problem

**Comment by coppolaemilio** [comment]
Thanks for the report, I'll look into it 👍🏻 
Have you tried downloading the repo itself (version 1.2.5) I made many bug fixes that *might* help?
I will try to reproduce it and fix it later today

**Comment by ababen1** [comment]
no I haven't tried that I just downloaded from the releases section, so i'll look into it ty
also, i forgot to mention it in the post but i changed the theme to have my project's native buttons and my project's font and made the dialog box smaller. could that have caused all the errors somehow?

**Comment by coppolaemilio** [comment]
> i changed the theme to have my project's native buttons and my project's font and made the dialog box smaller. could that have caused all the errors somehow?

Changing the theme or the button shouldn't affect the definitions at all

**Comment by ababen1** [comment]
> > i changed the theme to have my project's native buttons and my project's font and made the dialog box smaller. could that have caused all the errors somehow?
> 
> Changing the theme or the button shouldn't affect the definitions at all
ok then, i tried 1.2.5 and i couldn't even get it to start, it says it failed to retrieve the singleton, even though i checked and the path is correct 
![image](https://user-images.githubusercontent.com/80917990/124775409-451afa00-df47-11eb-9276-c42f000c9d4d.png)
![image](https://user-images.githubusercontent.com/80917990/124775520-5c59e780-df47-11eb-8083-3254baadb02d.png)

**Comment by Jowan-Spooner** [comment]
Make sure you rebooted godot at least once after installing

**Comment by ababen1** [comment]
i did

**Comment by coppolaemilio** [comment]
Pushing a fix soon to this issue

**Comment by coppolaemilio** [comment]
@ababen1 Made the changes, can you test it out again with the latest version? https://github.com/coppolaemilio/dialogic/commit/c1e04721bec80f1636f6eb9ee1d4d6162b590dcf

**Comment by ababen1** [comment]
> @ababen1 Made the changes, can you test it out again with the latest version? [c1e0472](https://github.com/coppolaemilio/dialogic/commit/c1e04721bec80f1636f6eb9ee1d4d6162b590dcf)

that fixed the singleton problem, but the problem with the definition is still the same, i tried to set a value and it crashed with this error:
![image](https://user-images.githubusercontent.com/80917990/124821079-f08f7300-df76-11eb-9282-273bd6007373.png)

**Comment by coppolaemilio** [comment]
@ababen1 pushed a new batch of changes (https://github.com/coppolaemilio/dialogic/commit/93176a4bbc194435e39a8d873b0c20f83e3fb965) to hopefully fix this singleton mess we have here. I will most likely overhaul the saving system since it is giving us a lot of headaches lately.

**Comment by ababen1** [comment]
> @ababen1 pushed a new batch of changes ([93176a4](https://github.com/coppolaemilio/dialogic/commit/93176a4bbc194435e39a8d873b0c20f83e3fb965)) to hopefully fix this singleton mess we have here. I will most likely overhaul the saving system since it is giving us a lot of headaches lately.

tried it, same error :(

**Comment by ababen1** [comment]
@coppolaemilio i've looked into the code and "reset saves" is true by default, i think that's what caused all the errors, it just deleted all the definitions i've entered in the editor when i ran the timeline 
i set it to false and it didnt crash and it showed the value correctly 
![image](https://user-images.githubusercontent.com/80917990/125481748-46593e42-bf85-4c2b-b036-9c02eff335ca.png)
may i ask what's the idea behind the "reset_saves" thing? i dont really understand what's the point of resetting my definitions at all, let alone doing it by default

**Comment by Jowan-Spooner** [comment]
Well I think in many smaller games you want the definitions to be reset at least in the beginning of the game. Also it can be confusing if you don't know they are saved when you quit your game. So when you test your game again it suddenly plays out differently. But yeah. It's a pretty confusing thing. I think your problem is a duplicate of #267 but maybe I'm mistaken.

**Comment by coppolaemilio** [comment]
Even if dialogic is resetting saves by default, why would it cause it to show those errors?

**Comment by ababen1** [comment]
> Even if dialogic is resetting saves by default, why would it cause it to show those errors?
i guess because it's trying to access data that was deleted

**Comment by coppolaemilio** [comment]
When the reset happen the data should go back to the defined value set on the Dialogic editor, but maybe you were calling the variables before the instance of Dialogic was created and able to read it?

**Comment by ababen1** [comment]
> When the reset happen the data should go back to the defined value set on the Dialogic editor, but maybe you were calling the variables before the instance of Dialogic was created and able to read it?

I don't know, i really didn't do anything complicated. just entered a variable in the editor, used brackets to display it in a dialogue, used dialogic.start() and added it as a child, and boom the definitions file is empty. i didnt even try to change the value in runtime or anything. just tried to display it

**Comment by RedXGames** [comment]
Hello @coppolaemilio , I found aWorkaround to this.
I think What's happening is that Definitions are only being created when you reset the definitions.

Basically All you need to do is call any dialogue at the start of the game with Dialogic.start(your_dialogue, true) # you have to reset defs
After that if you call the dialogue that modifies the definition, you won't get the Error above.
Of course this is not a fix as reseting definions at the game start will make saving them pointless but i think this could help you track the origin of the error.

So here's a final example to make it clear.
In my own game every dialogue that I create is called with Dialogic.start with reset_defs = false. Which means any definitoin that I create will give me the error above.
But if i at the start of the game reset the definitions, I wont get the error.

**Comment by RedXGames** [comment]
So I looked into the code and it seems in fact that if you never reset the defs at least once (Not every time you run the game, just when you add new definitions). The definitions wont get initiated and wont be saved in the config file.

This is the function code of Dialog.gd:
```
func load_config_files():
	if not Engine.is_editor_hint():
		if reset_saves:
			DialogicUtil.get_singleton('DialogicSingleton', self).init(reset_saves)
		definitions = DialogicUtil.get_singleton('DialogicSingleton', self).get_definitions()
	else:
		definitions = DialogicResources.get_default_definitions()
	settings = DialogicResources.get_settings_config()
	var theme_file = 'res://addons/dialogic/Editor/ThemeEditor/default-theme.cfg'
	if settings.has_section('theme'):
		theme_file = settings.get_value('theme', 'default')
	current_theme = load_theme(theme_file)
```

The "DialogicUtil.get_singleton('DialogicSingleton', self).init(reset_saves)" is only called when reset_saves is true.

I wish I can fix it myself but I dont know how.
Basically what needs to happen is for the plugin to detect when we add a new definition from the editor so it modifies the config file where the Definitions are saved so it can update them without resetting the previous ones.

**Comment by Jowan-Spooner** [comment]
I do believe what you describe is a duplicate of #267 but what's confusing on this problem is, that reset_saves is true by default in the Dialogic.start() function. Also the error in the editor should not happen as it should always try to load the default definitions anyways. But maybe these problems are connected, idk.

**Comment by RedXGames** [comment]
Exactly, It is true by default but  for it to work you have to assume that the function will be called with default parameters at least once, and in the DialogicSingleton , reset is false by default.
In my own project I never call the Dialogic.start with reset = true at any point of the game which lead to the same error.
I have tried it with multiple projects with different nodes structure and the result is always the same.

I managed to find a workaround for my main project by detecting if the player is starting a new game or loading an existing one : 
```
# when the player presses "New game button"
if new_game:
   DialogicSingleton.init(true)

# When I want to save the game data I also save the Dialogic definitions In my own save file: 
save_dict["definitions"] = DialogicSingleton.get_definitions()

# When The player presses The continue button I load the definitons : 
var defs = GameData.read_dict(save_dict, "definitions", {})
if "variables" in defs.keys():
for dict in defs["variables"]:
	Dialogic.set_variable(dict["name"], dict["value"])
```

Maybe this could help the OP with his game until the bug is fixed.

**Comment by coppolaemilio** [comment]
Since a new variable system is going to be implemented I'm closing this issue
